compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1348,33 +1348,47 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         span: Span,
         remaining_fields: FxHashMap<Ident, (usize, &ty::FieldDef)>,
     ) {
-        let tcx = self.tcx;
         let len = remaining_fields.len();
 
         let mut displayable_field_names =
             remaining_fields.keys().map(|ident| ident.as_str()).collect::<Vec<_>>();
 
         displayable_field_names.sort();
 
-        let truncated_fields_error = if len <= 3 {
-            String::new()
+        let truncated_fields_error: String;
+        let remaining_fields_names = if len <= 3 {
+            truncated_fields_error = String::new();
+
+            match len {
+                1 => format!("`{}`", displayable_field_names[0]),
+                2 => {
+                    format!("`{}` and `{}`", displayable_field_names[0], displayable_field_names[1])
+                }
+                3 => format!(
+                    "`{}`, `{}` and `{}`",
+                    displayable_field_names[0],
+                    displayable_field_names[1],
+                    displayable_field_names[2]
+                ),
+                _ => unreachable!(),
```

Comment:
I think you could use a `match` instead, for both `len <= 3` and `len > 3` cases.
```rust
match &remaining_fields[..] {
    [field1] => format!("`{}`", field1),
    [field1, field2] => ...,
    ...
}
```

Reply:
That looks a lot better!